File: problems/回溯/131.分割回文串.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;

bool isPalindrome(const string& s, int left, int right) {
    while (left < right) {
        if (s[left] != s[right]) return false;
        left++;
        right--;
    }
    return true;
}

void backtrack(vector<vector<string>>& result, vector<string>& path,
               const string& s, int start) {
    if (start == s.size()) {
        result.push_back(path);
        return;
    }
    for (int i = start; i < s.size(); i++) {
        if (isPalindrome(s, start, i)) {
            path.push_back(s.substr(start, i - start + 1));
            backtrack(result, path, s, i + 1);
            path.pop_back();
        }
    }
}

vector<vector<string>> partition(const string& s) {
    vector<vector<string>> result;
    vector<string> path;
    backtrack(result, path, s, 0);
    return result;
}

int main() {
    // 测试用例1
    vector<vector<string>> res1 = partition("aab");
    cout << "Test 1: s=\"aab\"" << endl;
    for (auto& partition : res1) {
        cout << "[";
        for (int i = 0; i < partition.size(); i++) {
            cout << "\"" << partition[i] << "\"";
            if (i < partition.size() - 1) cout << ",";
        }
        cout << "] ";
    }
    cout << endl;
    
    // 测试用例2
    vector<vector<string>> res2 = partition("a");
    cout << "Test 2: s=\"a\"" << endl;
    for (auto& partition : res2) {
        cout << "[";
        for (int i = 0; i < partition.size(); i++) {
            cout << "\"" << partition[i] << "\"";
            if (i < partition.size() - 1) cout << ",";
        }
        cout << "] ";
    }
    cout << endl;
    
    // 测试用例3
    vector<vector<string>> res3 = partition("abcba");
    cout << "Test 3: s=\"abcba\" - " << res3.size() << " partitions" << endl;
    
    return 0;
}

File: problems/回溯/131.分割回文串.py
#!/usr/bin/env python3
"""
131. 分割回文串 (Palindrome Partitioning)
题目: 给定一个字符串 s，将 s 分割成一些子串，使每个子串都是回文串。返回所有可能的分割方案。
算法: 回溯 + 中心扩展判断回文
时间复杂度: O(n * 2^n)
空间复杂度: O(n)
"""

def partition(s):
    result = []
    path = []
    
    def is_palindrome(sub):
        return sub == sub[::-1]
    
    def backtrack(start):
        if start == len(s):
            result.append(path[:])
            return
        for i in range(start, len(s)):
            sub = s[start:i + 1]
            if is_palindrome(sub):
                path.append(sub)
                backtrack(i + 1)
                path.pop()
    
    backtrack(0)
    return result

if __name__ == "__main__":
    # 测试用例1
    print(partition("aab"))
    # 输出: [["a","a","b"], ["aa","b"]]
    
    # 测试用例2
    print(partition("a"))
    # 输出: [["a"]]
    
    # 测试用例3
    print(partition("abcba"))
    # 输出: 所有回文分割方案
